### PR TITLE
[owic] fix malloc alignement

### PR DESCRIPTION
**Pull request by epatrizio**

### CODE REVIEW

**Comment by krtab** [comment]
Approved!

**Comment by krtab** [comment]
Actually @chambart spotted an issue with sizes below 16 that are not power of 2, don't merge yet

File: src/libc/src/stdlib.c [hunk 1]
```diff
@@ -24,24 +24,26 @@ extern unsigned char __heap_base;
...
-  /*   *((unsigned char *)bump_pointer + i) = 'i'; */
-  bump_pointer += size;
-  return (void *)owi_malloc(r, size);
+  unsigned int start;
+  unsigned int align = (size <= 16) ? size : 16;
+  unsigned int off_align = bump_pointer % align;
+  if ( off_align == 0 )
+    start = bump_pointer;
+  else
+    start = bump_pointer + (align - off_align);
```
Comment (zapashcanon) [inline src/libc/src/stdlib.c:33]: (here too)

**Review by krtab — APPROVED** [review]

**Comment by krtab** [comment]
(We nearly added back a commit from Leo that had been deleted on master 😅)

**Comment by filipeom** [comment]
> (We nearly added back a commit from Leo that had been deleted on master 😅)

Yeah this was my bad, probably should never do that again :sweat_smile:

File: src/libc/src/stdlib.c [hunk 2]
```diff
@@ -24,24 +24,29 @@ extern unsigned char __heap_base;
 unsigned int bump_pointer = &__heap_base;
 
 void *malloc(size_t size) {
-  unsigned int r = bump_pointer;
-  /* for (int i = 0; i < size; ++i) */
-  /*   *((unsigned char *)bump_pointer + i) = 'i'; */
-  bump_pointer += size;
-  return (void *)owi_malloc(r, size);
+  unsigned int start;
+  unsigned int closest_pow2 = 1 << (sizeof(size_t) - (__builtin_clz(size) + 1));
```
[the remark this answers is not in the document]
Comment (krtab) [reply src/libc/src/stdlib.c:28]: Yes my bad

File: src/libc/src/stdlib.c [hunk 3]
```diff
@@ -24,24 +24,29 @@ extern unsigned char __heap_base;
 unsigned int bump_pointer = &__heap_base;
 
 void *malloc(size_t size) {
-  unsigned int r = bump_pointer;
-  /* for (int i = 0; i < size; ++i) */
-  /*   *((unsigned char *)bump_pointer + i) = 'i'; */
-  bump_pointer += size;
-  return (void *)owi_malloc(r, size);
+  unsigned int start;
```
Comment (zapashcanon) [inline src/libc/src/stdlib.c:27]: `unsigned int start = bump_pointer;`

File: src/libc/src/stdlib.c [hunk 4]
```diff
@@ -24,24 +24,29 @@ extern unsigned char __heap_base;
...
+  unsigned int closest_pow2 = 1 << (sizeof(size_t)*8 - (__builtin_clz(size) + 1));
+  unsigned int align = (closest_pow2 <= 16) ? closest_pow2 : 16;
+  unsigned int off_align = bump_pointer % align;
+  if ( off_align == 0 ) {
+    start = bump_pointer;
+  }
```
Comment (zapashcanon) [inline src/libc/src/stdlib.c:33]: You can actually remove this `if`.

File: src/libc/src/stdlib.c [hunk 5]
```diff
@@ -24,24 +24,29 @@ extern unsigned char __heap_base;
...
+  else {
+    start = bump_pointer + (align - off_align);
```
Comment (zapashcanon) [inline src/libc/src/stdlib.c:35]: if (off_align != 0) {
  start += align - off_align;
}

Comment (krtab) [reply src/libc/src/stdlib.c:35]: (fwiw it makes no difference in the generated wasm)

**Comment by zapashcanon** [comment]
Thanks. I rewrote the code directly in 4a20b2ba3aa5217d69901903fabc368672bda6e0

**Pull request merged**